cpp/include/cuml/experimental/fil/detail/decision_forest_builder.hpp

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+#include <stdint.h>
+#include <numeric>
+#include <optional>
+#include <vector>
+#include <cuml/experimental/fil/postproc_ops.hpp>
+#include <cuml/experimental/fil/detail/forest.hpp>
+#include <cuml/experimental/fil/detail/index_type.hpp>
+#include <cuml/experimental/fil/exceptions.hpp>
+#include <cuml/experimental/fil/detail/raft_proto/buffer.hpp>
+#include <cuml/experimental/fil/detail/bitset.hpp>
+#include <cuml/experimental/fil/detail/raft_proto/ceildiv.hpp>
+#include <cuml/experimental/fil/detail/raft_proto/cuda_stream.hpp>
+#include <cuml/experimental/fil/detail/raft_proto/device_type.hpp>
+
+namespace ML {
+namespace experimental {
+namespace fil {
+namespace detail {
+
+/*
+ * Exception indicating that FIL model could not be built from given input
+ */
+struct model_builder_error : std::exception {
+  model_builder_error() : model_builder_error("Error while building model") {}
+  model_builder_error(char const* msg) : msg_{msg} {}
+  virtual char const* what() const noexcept { return msg_; }
+
+ private:
+  char const* msg_;
+};
+
+/*
+ * Struct used to build FIL forests
+ */
+template<typename decision_forest_t>
+struct decision_forest_builder {
+
+  /* The type for nodes in the given decision_forest type */
+  using node_type = typename decision_forest_t::node_type;
+
+  /* Add a root node, indicating the beginning of a new tree */
+  void start_new_tree() {
+    if (root_node_indexes_.size() == index_type{}) {
+      root_node_indexes_.emplace_back();
+    } else {
+      max_tree_size_ = std::max(cur_tree_size_, max_tree_size_);
+      if (alignment_ != index_type{}) {
+        if (cur_tree_size_ % alignment_ != index_type{}) {
+          auto padding = (alignment_ - cur_tree_size_ % alignment_);
+          for (auto i = index_type{}; i < padding; ++i) {
+            add_node(typename node_type::threshold_type{});
+          }
+        }
+      }
+      root_node_indexes_.push_back(
+        root_node_indexes_.back() + cur_tree_size_
+      );
+      cur_tree_size_ = index_type{};
```

Comment:
Yep! I prefer that spelling because I once worked in a codebase where we were using a ring as an index type where the conversion from integer 0 was not the zero-element of the ring. That's an unusual use case, but in general letting index types define their own zero without respect to another type seems useful. Even if we wanted to be more explicit, `index_type{0}` would be the way to go to stick with universal initialization syntax.